### DISCUSSION TITLE
Using GPUOcelot to emulate CUDA code on CPU

### DISCUSSION THREAD

**Discussion by mrdaybird** [gpuocelot/gpuocelot#5] [Q&A] [answered]
Is there any example on how to use gpuocelot to emulate a cuda file on the cpu(x86)?

I tried to use set `LD_PRELOAD` to the libgpuocelot.so and then compile the .cu file using nvcc (with cudart=shared). Though the compilation succeeded, there was segmentation fault at the runtime. It was a simple hello world program, so no problems there i guess.

**Comment by dmikushin** [comment, marked as answer]
Hi @mrdaybird , thank you for testing. GPUOcelot comes with a test called `gpuocelot_executive_TestEmulator`, which does the same. Could you please first check it works without a crash for you? If it works, please try to assign CMAKE_CUDA_ARCHITECTURES of your example code to 35-virtual (GPUOcelot does not support the most recent versions of PTX), like this:

```cmake
set(CMAKE_CUDA_ARCHITECTURES 35-virtual)
```

**Comment by mrdaybird** [comment]
Thank you! I think the setting the cuda architecture did the trick. 
Is the version 35(3.5?), the latest version of the cuda architecture supported by gpuocelot?

**Reply by dmikushin** [reply]
> Is the version 35(3.5?), the latest version of the cuda architecture supported by gpuocelot?

Yes. But I would like to eliminate a possible misunderstanding by noting the following:

You can still run the code compiled for sm_35 on newer generations of GPUs. This code will run, it just won't be able to use newer features implemented in more recent GPU hardware. It's the same as running SSE4 CPU code on a chip which supports AVX512.

**Comment by dmikushin** [comment]
Strictly speaking, GPUOcelot is sensitive to the PTX version, not to the compute capability. So it depends on the features that you use. If the test is super-basic, I bet you could even succeed with the most recent compute capability. The only requirement is: make sure the compute capability of your choice is **virtual**, meaning that the PTX code is included into your executable.

**Comment by hellodword** [comment]
I'm trying the GPUOcelot, with cuda 11.8 which maybe the last version that supports `compute_35`:

```
$ cat /etc/issue
Ubuntu 22.04.4 LTS \n \l

$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:33:58_PDT_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0
```

The `hello.cu`:

```cpp
extern "C" __global__ void hello_world(int N) {}
```

The `loader.cpp`:

```cpp
#include <fstream>
#include <math.h>

#include <cuda.h>

#define CHK(X)                                           \
  if ((err = X) != CUDA_SUCCESS)                         \
    printf("CUDA error %d at %d\n", (int)err, __LINE__); \
  if (err != CUDA_SUCCESS)                               \
  return 1

int main()
{
  CUresult err;

  CHK(cuInit(0));

  int device_count;
  CHK(cuDeviceGetCount(&device_count));
  printf("device count %d\n", device_count);

  CUdevice cuDevice;
  CHK(cuDeviceGet(&cuDevice, 0));

  char device_name[255];
  CHK(cuDeviceGetName(device_name, 255, cuDevice));
  printf("device name %s\n", device_name);

  CUcontext cuContext;
  CHK(cuCtxCreate(&cuContext, 0, cuDevice));

  std::ifstream file("hello.ptx");
  std::string ptxContent((std::istreambuf_iterator<char>(file)),
                         std::istreambuf_iterator<char>());
  file.close();

  CUmodule cuModule;
  CHK(cuModuleLoadData(&cuModule, ptxContent.c_str()));

  CUfunction cuFunction;
  CHK(cuModuleGetFunction(&cuFunction, cuModule, "hello_world"));

  int N = 1;
  void *args[] = {&N};
  CHK(cuLaunchKernel(cuFunction,
                     1, 1, 1,
                     1, 1, 1,
                     0, NULL,
                     args, NULL));

  CHK(cuCtxSynchronize());
  CHK(cuModuleUnload(cuModule));
  CHK(cuCtxDestroy(cuContext));

  return 0;
}

```

Then I generate the `hello.ptx` with `nvcc -arch=compute_35 -ptx -o hello.ptx ./hello.cu`:

```
//
// Generated by NVIDIA NVVM Compiler
//
// Compiler Build ID: CL-
// Cuda compilation tools, release 11.8, V11.8.89
// Based on NVVM 7.0.1
//

.version 7.8
.target sm_35
.address_size 64

	// .globl	hello_world

.visible .entry hello_world(
	.param .u32 hello_world_param_0
)
{



	ret;

}
```

Build and run:

```
$ g++ hello_ptx_loader.cpp -o hello -I/usr/local/cuda/include -L/usr/local/cuda/lib64 -lcuda
$ LD_PRELOAD=/path/to/libgpuocelot.so  ./hello
 - cuInit() 
==Ocelot== WARNING: Failed to find 'configure.ocelot' in current directory, loading defaults.

==Ocelot== INFO: You may consider adding one if you need to change the Ocelot target, or runtime options.

 - cuDeviceGetCount() 
device count 1
 - cuDeviceGet() 
 - cuDeviceGetName() 
device name Ocelot PTX Emulator
 - cuCtxCreate_v2() 
 - cuModuleLoadData() 
 - cuModuleGetFunction() 
CUDA error 3 at 45
```

The `cuLaunchKernel` always returns `CUDA_ERROR_NOT_INITIALIZED` even the `cuInit` returns `CUDA_SUCCESS`.

Any idea?

**Reply by dmikushin** [reply]
Hi @hellodword ,

> cuLaunchKernel always returns CUDA_ERROR_NOT_INITIALIZED

This is an indication of `cuLaunchKernel` being taken from the real libcuda.so instead of the gpuocelot's preloaded library. Why it could happen? I recall there are two different versions of kernel launching function, and it could be that only one is implemented in gpuocelot, not the other one. Could you please check?

**Reply by hellodword** [reply]
~~I tried [`cuLaunch` and `cuLaunchGrid`](https://github.com/gpuocelot/gpuocelot/blob/c476bcb605579b123b85d34afef2c848127e8cdc/ocelot/src/cuda/CudaDriverApi.cpp#L742-L751) but no luck, got the same `CUDA_ERROR_NOT_INITIALIZED`~~

**Reply by dmikushin** [reply]
Thank you for testing, let me look at this

**Reply by hellodword** [reply]
Sorry my mistake, the `cuLaunchGrid` works, with `cuParamSetv` :)

**Reply by hellodword** [reply]
I'm following the [cuda/test/driver/generic.cpp](https://github.com/gpuocelot/gpuocelot/blob/c476bcb605579b123b85d34afef2c848127e8cdc/ocelot/src/cuda/test/driver/generic.cpp), it helps a lot, thanks

**Comment by hellodword** [comment]
Replacing `cuLaunchKernel` with `cuLaunchGrid` solved the issue above, but the `cuCtxDestroy` ran into the same issue.

To reproduce:

```diff
diff --git a/ocelot/src/cuda/test/driver/sequence.cpp b/ocelot/src/cuda/test/driver/sequence.cpp
index 4b33397..0aae904 100644
--- a/ocelot/src/cuda/test/driver/sequence.cpp
+++ b/ocelot/src/cuda/test/driver/sequence.cpp
@@ -3,8 +3,8 @@
        \brief CUDA Driver API version of the sequence program
 */
 
-#include <ocelot/cuda/cuda.h>
-//#include <cuda.h>
+// #include <ocelot/cuda/cuda.h>
+#include <cuda.h>
 
 #include <iostream>
 
@@ -154,7 +154,11 @@ int main() {
        }
        
        cuModuleUnload(module);
-       cuCtxDestroy(ctx);
+       result = cuCtxDestroy(ctx);
+       if (result != CUDA_SUCCESS) {
+               report("cuCtxDestroy() failed: " << result);
+               // return 1;
+       }
        
        int errors = 0;
        for (int i = 0; i < N; i++) {
```

Everything works great but the `cuCtxDestroy` returns `CUDA_ERROR_NOT_INITIALIZED`

```
$ LD_PRELOAD=/path/to/libgpuocelot.so  ./sequence 
 - cuInit() 
==Ocelot== WARNING: Failed to find 'configure.ocelot' in current directory, loading defaults.

==Ocelot== INFO: You may consider adding one if you need to change the Ocelot target, or runtime options.

 - cuDriverGetVersion() 
cuda driver version: 12290
 - cuDeviceGetCount() 
  1 devices
 - cuDeviceGet() 
 - cuDeviceGetName() 
using device 'Ocelot PTX Emulator'
 - cuDeviceComputeCapability() 
  compute capability 2.1
 - cuCtxCreate_v2() 
 - cuDeviceGetAttribute() 
  compute mode: 1
 - cuModuleLoad() 
 - cuModuleGetFunction() 
 - cuMemAlloc_v2() 
 - cuMemcpyHtoD_v2() 
 - cuParamSetSize() 
 - cuParamSetv() 
 - cuParamSetv() 
 - cuFuncSetBlockShape() 
 - cuLaunchGrid() 
 - cuMemcpyDtoH_v2() 
 - cuModuleUnload() 
cuCtxDestroy() failed: CUDA ERROR: 3
pass/fail: Passed
```

**Reply by dmikushin** [reply]
@hellodword , please explain your use case in more detail. I'm getting puzzled by what you are trying to archive by doing all these steps.

1. You create an OpenCL-style PTX code loader, OK.
2. You compile the loader to load your PTX, but you link it with libcuda:
```
g++ hello_ptx_loader.cpp -o hello -I/usr/local/cuda/include -L/usr/local/cuda/lib64 -lcuda
```
3. So, this loader is based on the CUDA mainline. But then suddently you change your mind and want to run the loader through the GPUOcelot:
```
LD_PRELOAD=/path/to/libgpuocelot.so  ./hello
```
I'm wondering what is this for. If you want to load PTX by GPUOcelot, then please link libgpuocelot from the very beginning instead of libcuda. Why you might need to first link libcuda and then switch to libgpuocelot? I think the purpose of such experiment is indeed to reveal incompatibilities between libcuda and libgpuocelot and get failures because of that. You seem to be unsatisfied with the outcome, though the issues that you get are logical to happen. In order to help you, I need to understand what is your goal.

**Reply by hellodword** [reply]
@dmikushin , my goal is running cuda code on CPU. And I learned the `LD_PRELOAD` thing from [here](https://github.com/gpuocelot/gpuocelot/discussions/13), I thought it's the way to use GPUOcelot.

I'll try to link the libgpuocelot directly, thanks a lot.

**Reply by hellodword** [reply]
With a small patch, it works:

```diff
diff --git a/ocelot/src/cuda/CudaDriverApi.cpp b/ocelot/src/cuda/CudaDriverApi.cpp
index a9aea07..925572c 100644
--- a/ocelot/src/cuda/CudaDriverApi.cpp
+++ b/ocelot/src/cuda/CudaDriverApi.cpp
@@ -167,6 +167,11 @@ CUresult Function(cuCtxDestroy)( CUcontext ctx ) {
        RETURN( CudaApi::get()->cuCtxDestroy(ctx) );
 }
 
+CUresult FunctionV(cuCtxDestroy)( CUcontext ctx ) {
+       trace();
+       RETURN( CudaApi::get()->cuCtxDestroy(ctx) );
+}
+
 CUresult Function(cuCtxAttach)(CUcontext *pctx, unsigned int flags) {
        trace();        
        RETURN( CudaApi::get()->cuCtxAttach(pctx, flags) );
```

[The `#define cuCtxDestroy  cuCtxDestroy_v2` is defined in `cuda.h` of cuda 11.8](https://gitlab.com/nvidia/headers/cuda-individual/cudart/-/blob/cuda-11.8.0/cuda.h#L130), but ocelot does not hijack the `cuCtxDestroy_v2`, so it fallbacks to the libcuda:

```
$ objdump -T ./libgpuocelot.so | awk '{print $7}' | sort -u | grep '^cu[A-Z]' | grep -i ctx
cuCtxAttach
cuCtxCreate
cuCtxCreate_v2
cuCtxDestroy
cuCtxDetach
cuCtxGetDevice
cuCtxPopCurrent
cuCtxPushCurrent
cuCtxSynchronize
cuGLCtxCreate
```

@dmikushin How do you think about this patch? Could I send this as a pull request :)

**Reply by dmikushin** [reply]
Yes, exactly. I meant to say that CUDA Driver API is far less stable than the CUDA Runtime API. Therefore, it's a higher chance that the libgpuocelot and libcuda may diverge, which you have encountered. But in CUDA Runtime API it is unlikely to happen. So, the general recommendation is not to hijack the driver API, but hijack the runtime API when needed by using LD_PRELOAD. So, use LD_PRELOAD with a normal CUDA program. But for low-level API, as you do, instead of hijacking the driver, easier is to link against libgpuocelot directly. Anyway, great that you have sorted this out yourself!

> Could I send this as a pull request

Yes, the PR is most welcome!

**Reply by dmikushin** [reply]
A generic solution to the driver API divergence in the next version of GPUOcelot might be to examine all cu* symbols of the application and create wrappers for all of them, including the possible suffix variances that they may have.
